### PR TITLE
solve the problem of zero sample

### PR DESCRIPTION
**Pull request by DavideTisi**
solve #158 

```python
block = TensorBlock(
            values=np.zeros([0, 1]),
            properties=Labels(["prop"], np.zeros([1, 1], dtype=int)),
            samples=Labels(["structure"], np.empty((0, 1), dtype=np.int32)),
            components=[],
        )

result_block = TensorBlock(
            values=np.zeros([0, 1]),
            properties=Labels(["prop"], np.zeros([1, 1], dtype=int)),
            samples=Labels([], np.empty((0, 0), dtype=np.int32)),
            components=[],
        )

tensor = TensorMap(Labels.single(), [block])
result_tensor = TensorMap(Labels.single(), [result_block])

tensor_sum = equistore.sum_over_samples(tensor, "structure")

self.assertTrue(equistore.equal(result_tensor, tensor_sum))
```

or for gradient 

```python
block = TensorBlock(
            values=np.array(
                [[1, 2, 4], [3, 5, 6], [-1.3, 26.7, 4.54], [3.5, 5.3, 6.87]]
            ),
            samples=Labels(
                ["structure", "samples1"],
                np.array(
                    [
                        [0, 0],
                        [0, 1],
                        [1, 0],
                        [1, 1],
                    ],
                    dtype=np.int32,
                ),
            ),
            components=[],
            properties=Labels(
                ["properties"], np.array([[0], [1], [5]], dtype=np.int32)
            ),
        )

block.add_gradient(
            parameter="positions",
            data=np.zeros((0, 3)),
            samples=Labels(["sample", "structure"], np.empty((0, 2), dtype=np.int32)),
            components=[],
        )

sum_block = TensorBlock(
            values=np.array([[-0.3, 28.7, 8.54], [6.5, 10.3, 12.87]]),
            samples=Labels(
                ["samples1"],
                np.array(
                    [[0], [1]],
                    dtype=np.int32,
                ),
            ),
            components=[],
            properties=Labels(
                ["properties"], np.array([[0], [1], [5]], dtype=np.int32)
            ),
        )

sum_block.add_gradient(
            parameter="positions",
            data=np.zeros((0, 3)),
            samples=Labels(["sample", "structure"], np.empty((0, 2), dtype=np.int32)),
            components=[],
        )

tensor = TensorMap(Labels.single(), [block])
tensor_sum_result = TensorMap(Labels.single(), [sum_block])

tensor_sum = equistore.sum_over_samples(tensor, "structure")
self.assertTrue(equistore.allclose(tensor_sum_result, tensor_sum, atol=1e-14))
```

----
:books: Documentation preview :books:: https://equistore--164.org.readthedocs.build/en/164/

### CODE REVIEW

File: python/tests/operations/reduce_over_samples.py [hunk 1]
```diff
@@ -941,6 +941,136 @@ def test_reduction_of_one_element(self):
...
+        self.assertTrue(equistore.equal(result_tensor, tensor_variance))
+        self.assertTrue(equistore.equal(result_tensor, tensor_sqrt))
+
+    def test_zeros_sample_block_gradient(self):
+        # block = TensorBlock(
+        #     values=np.zeros([2, 1]),
+        #     properties=Labels(["prop"], np.zeros([1, 1], dtype=int)),
+        #     samples=Labels(["structure"], np.array([[0], [1]])),
+        #     components=[],
+        # )
```
Comment (PicoCentauri) [inline python/tests/operations/reduce_over_samples.py:1006]: Remove?

File: python/src/equistore/operations/reduce_over_samples.py [hunk 2]
```diff
@@ -28,11 +28,43 @@ def _reduce_over_samples_block(
...
+        # Here is different from the general case where we use Labels.single() if
+        # if len(remaining_samples) == 0
+        # Labels.single() cannot be used because Labels.single() has not
+        # an np.empty() array as values but has one values, it has dimension (1,...)
+        # we want (0,...).
+        # here if len(remaining_samples) == 0 -> Labels([], shape=(0, 0), dtype=int32)
+
+        samples_label = Labels(
+            remaining_samples,
+            np.empty((0, len(remaining_samples)), dtype=np.int32),
```
Comment (PicoCentauri) [inline python/src/equistore/operations/reduce_over_samples.py:47]: I think zeros is good enough here.

```suggestion
            np.zeros((0, len(remaining_samples)), dtype=np.int32),
```

File: python/src/equistore/operations/reduce_over_samples.py [hunk 3]
```diff
@@ -93,6 +125,22 @@ def _reduce_over_samples_block(
...
 
     for parameter, gradient in block.gradients():
+        # check if all gradients are zeros
+        if gradient.samples.shape[0] == 0:
+            # The Gradient does not change because, if they are all zeros, also the
+            # gradients of a reduce operation is zero.
+            # For any function of the TensorBlock values x(t):
+            # f(x(t))-> df(x(t))/dx * dx/dt
+            # and dx/dt == 0.
+            for parameter, gradient in block.gradients():
```
Comment (PicoCentauri) [inline python/src/equistore/operations/reduce_over_samples.py:135]: Why do we have to do a double loop here? You are already looping over the gradients.

**Review by PicoCentauri — CHANGES_REQUESTED** [review]
Very nice! The code for the reduce over sample operations also starts to become quite complex...

File: python/src/equistore/operations/reduce_over_samples.py (hunk 3 above)
Comment (DavideTisi) [reply python/src/equistore/operations/reduce_over_samples.py:135]: because of copy-paste

File: python/tests/operations/reduce_over_samples.py [hunk 4]
```diff
@@ -941,6 +941,136 @@ def test_reduction_of_one_element(self):
...
+            samples=Labels([], np.empty((0, 0), dtype=np.int32)),
+            components=[],
+        )
+
+        tensor = TensorMap(Labels.single(), [block])
+        result_tensor = TensorMap(Labels.single(), [result_block])
+
+        tensor_sum = equistore.sum_over_samples(tensor, "structure")
+        tensor_mean = equistore.mean_over_samples(tensor, "structure")
+        tensor_sqrt = equistore.std_over_samples(tensor, "structure")
```
Comment (Luthaf) [inline python/tests/operations/reduce_over_samples.py:965]:
```suggestion
        tensor_std = equistore.std_over_samples(tensor, "structure")
```

File: python/tests/operations/reduce_over_samples.py [hunk 5]
```diff
@@ -941,6 +941,136 @@ def test_reduction_of_one_element(self):
...
+            samples=Labels(["s1"], np.empty((0, 1), dtype=np.int32)),
+            components=[],
+        )
+
+        tensor = TensorMap(Labels.single(), [block])
+        result_tensor = TensorMap(Labels.single(), [result_block])
+
+        tensor_sum = equistore.sum_over_samples(tensor, "structure")
+        tensor_mean = equistore.mean_over_samples(tensor, "structure")
+        tensor_sqrt = equistore.std_over_samples(tensor, "structure")
```
Comment (Luthaf) [inline python/tests/operations/reduce_over_samples.py:992]:
```suggestion
        tensor_std = equistore.std_over_samples(tensor, "structure")
```

File: python/tests/operations/reduce_over_samples.py [hunk 6]
```diff
@@ -941,6 +941,136 @@ def test_reduction_of_one_element(self):
...
+            samples=Labels(["sample", "structure"], np.empty((0, 2), dtype=np.int32)),
+            components=[],
+        )
+
+        tensor = TensorMap(Labels.single(), [block])
+        tensor_sum_result = TensorMap(Labels.single(), [sum_block])
+
+        tensor_sum = equistore.sum_over_samples(tensor, "structure")
+        tensor_mean = equistore.mean_over_samples(tensor, "structure")
+        tensor_sqrt = equistore.std_over_samples(tensor, "structure")
```
Comment (Luthaf) [inline python/tests/operations/reduce_over_samples.py:1063]:
```suggestion
        tensor_std = equistore.std_over_samples(tensor, "structure")
```

File: python/tests/operations/reduce_over_samples.py [hunk 7]
```diff
@@ -941,6 +941,136 @@ def test_reduction_of_one_element(self):
...
+            components=[],
+            properties=Labels(
+                ["properties"], np.array([[0], [1], [5]], dtype=np.int32)
+            ),
+        )
+
+        block.add_gradient(
+            parameter="positions",
+            data=np.zeros((0, 3)),
+            samples=Labels(["sample", "structure"], np.empty((0, 2), dtype=np.int32)),
```
Comment (Luthaf) [inline python/tests/operations/reduce_over_samples.py:1032]:
```suggestion
            samples=Labels(["sample", "structure", "atom"], np.zero((0, 3), dtype=np.int32)),
```

File: python/tests/operations/reduce_over_samples.py [hunk 8]
```diff
@@ -941,6 +941,136 @@ def test_reduction_of_one_element(self):
...
+            components=[],
+            properties=Labels(
+                ["properties"], np.array([[0], [1], [5]], dtype=np.int32)
+            ),
+        )
+
+        sum_block.add_gradient(
+            parameter="positions",
+            data=np.zeros((0, 3)),
+            samples=Labels(["sample", "structure"], np.empty((0, 2), dtype=np.int32)),
```
Comment (Luthaf) [inline python/tests/operations/reduce_over_samples.py:1054]:
```suggestion
            samples=Labels(["sample", "structure", "atom"], np.empty((0, 3), dtype=np.int32)),
```

**Comment by DavideTisi** [comment]
> Very nice! The code for the reduce over sample operations also starts to become quite complex...

yes, it is becoming quite complex

**Review by PicoCentauri — APPROVED** [review]
Yes, looks like this will solve the issue.

**Pull request merged**